hackathon_site/dashboard/frontend/src/slices/hardware/hardwareSlice.ts

```diff
@@ -0,0 +1,44 @@
+import { createSlice, createEntityAdapter } from "@reduxjs/toolkit";
+
+export interface Hardware {
+    id: number;
+    name: string;
+    model_number: string;
+    manufacturer: string;
+    datasheet: string;
+    quantity_available: number;
+    notes: string;
+    max_per_team: number;
+    picture: string;
+    categories: number[];
+    quantity_remaining: string;
+}
+
+export interface HardwareState {
```

Comment:
This isn't actually the type of the hardware state, since it doesn't include the keys added by `createEntityAdapter`.

You could do

```typescript
// Note: not exported
interface HardwareExtraState {
    isLoading: boolean;
    error: string | null;
    next: string | null;
}

const extraState: HardwareExtraState = {
    isLoading: false,
    error: null,
    next: null,
};

const hardwareAdapter = createEntityAdapter<Hardware>();

export const initialState = hardwareAdapter.getInitialState(extraState);
export type HardwareState = typeof initialState;
```

With all that, we can create properly typed selectors and stuff. For example:

```typescript
export const isLoadingSelector = createSelector(
    hardwareSliceSelector,
    (state: HardwareState) => state.isLoading
);
```

If you did that, your IDE should be able to properly autocomplete the fields that `state` has (the ones in `HardwareExtraState`, as well as `ids` and `entities`).